content/blog/2019-12-03-posits/index.md

```diff
@@ -0,0 +1,262 @@
+
+Finally, we compute percent error by comparing `float` and `posit` benchmark results to the `double` baseline benchmark.
+We regard benchmark errors as arising from the inaccuracy that is characteristic of the particular 32-bit representation, compounded with successive operations.
+With our deepest apologies to Gustafson, we compute these errors on a linear scale, rather than a logarithmic one, and use these as metrics to compare the accuracy of the two representations.
+
+
+## LLVM pass implementation
+
+We aimed for our `float` and `posit` passes to insert as much representation-specific functionality as possible into our benchmarks.
+The case of `float`s allows a full translation, since every `double` operation can be cast as a `float` operations.
+The case of `posit`s lacked full support, however: we did not provide implementations for all existing LLVM `double` operations.
+
+To accommodate both cases under one pass implementation, 
+we designed our pass such that it first lowers each supported `double` operation and operands down to the target representation, subsequently lifting the result back to a `double`.
+
+Since each rewrite has a purely local effect, we can avoid reasoning about complex allocation and type casting scenarios.
+Though computationally inefficient, this approach immediately supports all programs and is easily extendible.
+
+In pseduo-LLVM, the general structure of a pass to type `typ` converts an arbitrary `double` operation `op` and its operands proceeds as follows:
+
+```llvm 
+x : double = ..
+y : double = ..
+result : double = op double x y
+```
+
+-->
+
+
+```llvm
+x : double = ..
+y : double = ..
+
+x_typ : typ = convert double x to typ;
+y_typ : typ = convert double y to typ;
+result_typ : typ = typ_op x_typ y_typ; 
+
+result double = convert result_typ typ to double
+```
+
+This generic pass structure parametrizes `typ` into three components:
+- `convert double to typ`: conversion from `double` to `typ`
+- `convert typ to double`: conversion from `typ` to `double`
+- `typ_op`: implementation over `typ` values of `op`, the corresponding `double` operation
+
+The `float` pass specifies these components as:
+- `fptrunc double to float`
+- `fpext float to double`
+- LLVM's floating point operations with type parameter set to `float` (e.g., `fadd float x y`)
+
+The `posit` pass draws these components from external `C` functions implemented in [Cerlane Leong's SoftPosit repository](https://gitlab.com/cerlane/SoftPosit-Python).
+In particular, we borrow the basic arithmetic operations (`+`, `-`, `*`, and `/`) over 32-bit posits. 
+
+
+### Benchmark Results
+
+We ran several benchmarks from [FPBench](https://fpbench.org/benchmarks.html) that accomodated our limited selection of posit operators.
+Here, we list their names, sample sizes, and associated results:
```

Comment:
I think you should explain a little more about what these results mean:

- What is "error"? What is it relative to?
- What is the distribution, i.e., what collection did you take the mean & standard deviation over?

Reply:
Clarified this: error = percent error = (correct - observed / correct), where correct = value of double benchmark.
Distribution of inputs is linearly sampled over the ranges specified for each FPBench benchmark.
Sample sizes scale with size of input spaces while remaining computationally feasible (heuristically, on the order of ~20^(num_inputs) )

Reply:
Nice; that helps a lot.